[Concept Entry] Luau: iterators.md (#3418)
Codecademy/docs: content/luau/concepts/iterators/iterators.md

### `repeat` loop example and syntax

```lua
repeat
	print("This will always run at least once.")
until false
```

This results in the output:

```shell
This will always run at least once.
```

## `for` loops

A `for` loop iterates through a series of elements and executes a code block at each one of those elements. This series could be a series of numbers, elements in a table, or characters in a string.

### Numeric for loops

A numeric for loop iterates through a series of numbers. A numeric for loop includes a start value, end value, and an optional increment value.

```pseudo
for i = start_value, end_value, increment_value do
	--put code here
end
```

`start_value` initializes the value of `i`. `increment_value` increments the value of `i` after each run of the loop. If `increment_value` is positive, the loop will end once `i` is greater than or equal to `end_value`. If `increment_value` is negative, the will end when `i` is less than or equal to `end_value`.

#### Numeric `for` loop examples

Excluding the optional increment value parameter results in a default increment value of `1`:

```lua
for i = 1, 3 do

> 	print(i)

Reviewer: be sure to use 2 spaces for code indentations not tabs

end
```

Output:

```shell
1
2
3
```

An example using a negative increment value:

```lua
for i = 10, 1, -2 do
	print(i)
end
```

Output:

```shell
10
8
6
4
2
```

### Generic `for` loops

A generic `for` loop iterates through the indices of a table or the characters in a string.

#### `pairs()`

The `pairs()` function can be used to iterate through all elements in a table:

```lua
mixed_table = {}
mixed_table[1] = 1
mixed_table[2] = 0.5
mixed_table[5] = "biff"
mixed_table["a"] = 7

for i,v in pairs(mixed_table) do
	print("The value at index " .. i .. " is " .. v)
end
```

Output:

```shell
The value at index 1 is 1
The value at index 2 is 0.5
The value at index 5 is biff
The value at index a is 7
```

> Note that the order in which the elements of the table are looped through is not guaranteed when using `pairs()`.

Reviewer: ```suggestion
> **Note:** The order in which the elements of the table are looped through is not guaranteed when using `pairs()`.
```

#### `ipairs()`

The `ipairs()` function can also be used to iterate through a table, but it has a few key differences from `pairs()`:

1. `ipairs()` will ignore any elements with non-integer indices.
2. `ipairs()` will begin at index `1` and ignore any elements at earlier indices.
3. `ipairs()` will stop iterating if there are any integer gaps in the indices.
4. `ipairs()` guarantees the order in which the elements are looped through.

Due to these differences, `ipairs()` is best used for iterating through arrays.

> Compare the behavior of `ipairs()` when used on an identical table as in the previous example:

Reviewer: ```suggestion
Compare the behavior of `ipairs()` when used on the same table as above:
```
